mm2src/common/patterns/state_machine.rs

```diff
@@ -0,0 +1,291 @@
+use async_trait::async_trait;
+
+pub mod prelude {
+    pub use super::{LastState, State, StateExt, StateMachine, StateResult, TransitionFrom};
+}
+
+pub struct StateMachine<Ctx, Result> {
+    /// The shared between states context.
+    ctx: Ctx,
+    phantom: std::marker::PhantomData<Result>,
+}
+
+impl<Ctx: Send + 'static, Result: 'static> StateMachine<Ctx, Result> {
+    pub fn from_ctx(ctx: Ctx) -> Self {
+        StateMachine {
+            ctx,
+            phantom: std::marker::PhantomData::default(),
+        }
+    }
+
+    pub async fn run(mut self, initial_state: impl State<Ctx = Ctx, Result = Result>) -> Result {
+        let mut state: Box<dyn State<Ctx = Ctx, Result = Result>> = Box::new(initial_state);
+        loop {
+            let result = state.on_changed(&mut self.ctx).await;
+            let next_state = match result {
+                StateResult::ChangeState(ChangeGuard { next }) => next,
+                StateResult::Finish(ResultGuard { result }) => return result,
+            };
+
+            state = next_state;
+        }
+    }
+}
+
+pub trait TransitionFrom<Prev> {}
+
+/// Prevent implementing [`TransitionFrom<T>`] for `Next` If `T` implements `LastState` already.
+impl<T, Next> !TransitionFrom<T> for Next
+where
+    T: LastState,
+    // this bound is required to prevent conflicting implementation with `impl<T> !TransitionFrom<T> for T`.
+    (T, Next): NotSame,
+{
+}
+
+/// Prevent implementing [`TransitionFrom<T>`] for itself.
+impl<T> !TransitionFrom<T> for T {}
+
+#[async_trait]
+pub trait State: Send + 'static {
+    type Ctx: Send;
+    type Result;
+
+    /// An action is called on entering this state.
+    /// To change the state to another one in the end of processing, use [`StateExt::change_state`].
+    /// For example:
+    /// ```rust
+    /// return Self::change_state(next_state);
+    /// ```
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result>;
+}
+
+pub trait StateExt {
+    /// Change the state to the `next_state`.
+    /// This function performs the compile-time validation whether this state can transition to the `Next` state,
+    /// i.e checks if `Next` implements [`Transition::from(ThisState)`].
+    fn change_state<Next>(next_state: Next) -> StateResult<Next::Ctx, Next::Result>
+    where
+        Self: Sized,
+        Next: State + TransitionFrom<Self>,
+    {
+        StateResult::ChangeState(ChangeGuard::next(next_state))
+    }
+}
+
+impl<T: State> StateExt for T {}
+
+#[async_trait]
+pub trait LastState: Send + 'static {
+    type Ctx: Send;
+    type Result;
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result;
+}
+
+#[async_trait]
+impl<T: LastState> State for T {
+    type Ctx = T::Ctx;
+    type Result = T::Result;
+
+    /// The last state always returns the result of the state machine calculations.
+    async fn on_changed(self: Box<Self>, ctx: &mut T::Ctx) -> StateResult<Self::Ctx, Self::Result> {
```

Comment:
Just wondering if we can avoid boxing here and in other places if present? It will be awesome if ever possible 🙂

Reply:
Unfortunately, I tried to avoid boxing but didn't find a way to do it. I'm not sure if I can explain it in a simple and understandable way :D
`StateResult` cannot depend on a `Next` type, because `State::on_changed` may return various next states (therefore various `Next` types), but Rust doesn't allow to do this. It's because `StateResult` contains `Box<dyn State>`.

Secondly, let's imagine that we have a trait like this:
```rust
trait State {
    fn on_changed(&self) -> StateResult;
}
let obj: Box<dyn State> = todo!();
// We can call the `State::on_changed` method on the `Box<dyn State>` object.
obj.on_changed();
```

Lets change the trait:
```rust
trait State {
    fn on_changed(self) -> StateResult;
    fn on_changed_boxed(self: Box<Self>) -> StateResult;
}
let obj: Box<dyn State> = todo!();
// We cannot call the `State::on_changed` method on the `Box<dyn State>` object.
// obj.on_changed();

// Only if it takes a boxed self.
obj.on_changed_boxed();
```